src/NServiceBus.Core/Sagas/SagaMessage.cs

```diff
@@ -10,16 +12,22 @@ public class SagaMessage
         /// </summary>
         /// <param name="messageType">Type of the message.</param>
         /// <param name="isAllowedToStart"><code>true</code> if the message can start the saga, <code>false</code> otherwise.</param>
-        public SagaMessage(string messageType, bool isAllowedToStart)
+        public SagaMessage(Type messageType, bool isAllowedToStart)
         {
             MessageType = messageType;
+            MessageTypeName = messageType.FullName;
             IsAllowedToStartSaga = isAllowedToStart;
         }
 
         /// <summary>
         /// The type of the message.
         /// </summary>
-        public string MessageType { get; private set; }
+        public Type MessageType { get; private set; }
```

Comment:
@danielmarbach: "Typeless sagas"? OK, I might give you that. But typeless *messages*??

Another option would be internal constructor, internal GetClrType() method.

```csharp
public class SagaMessage
{
    internal SagaMessage(Type messageType, bool isAllowedToStart);
    internal Type GetClrType();
    public string MessageType { get; private set; }
    public bool IsAllowedToStartSaga {get; private set; }
}
```

Although I am not sure exactly why the ctor is public right now. I don't see any immediate need for downstreams to access it but I need to check NH and Raven to see if they use it for tests or something.

The biggest cringe I have with that is that if we wanted to add type as public later, the fact that MessageType is string would make adding a public Type property later very awkward. What do you call it? MessageClrType?